components-api/src/main/scala/pl/touk/nussknacker/engine/api/typed/TypeConversionHandler.scala

```diff
@@ -74,14 +74,25 @@ object TypeConversionHandler {
     val boxedSuperclassCandidate = ClassUtils.primitiveToWrapper(superclassCandidate.klass)
     // We can't check precision here so we need to be loose here
     // TODO: Add feature flag: strictNumberPrecisionChecking (default false?)
-    if (NumberTypesPromotionStrategy
-        .isFloatingNumber(boxedSuperclassCandidate) || boxedSuperclassCandidate == classOf[java.math.BigDecimal]) {
-      ClassUtils.isAssignable(boxedGivenClass, classOf[Number], true)
-    } else if (NumberTypesPromotionStrategy.isDecimalNumber(boxedSuperclassCandidate)) {
-      ConversionFromClassesForDecimals.exists(ClassUtils.isAssignable(boxedGivenClass, _, true))
-    } else {
-      false
+
+    def isFloating(candidate: Class[_]): Boolean = {
+      NumberTypesPromotionStrategy.isFloatingNumber(candidate) || candidate == classOf[java.math.BigDecimal]
+    }
+    def isDecimalNumber(candidate: Class[_]): Boolean = {
+      NumberTypesPromotionStrategy.isDecimalNumber(candidate)
+    }
+
+    boxedSuperclassCandidate match {
+      case candidate if isFloating(candidate) =>
```

Comment:
Treating specially implicit conversion for floats is not a way to go, we should have option to check can be subclass without conversion in general

Reply:
To clarify: we already treat floats specially, the condition here is not changed: `if` is replaced with pattern matching, the condition is extracted to separate `def` above.

The only change here is in `if isDecimalNumber` section:
```
ConversionFromClassesForDecimals.exists(ClassUtils.isAssignable(boxedGivenClass, _, true))
```
that checks `boxedGivenClass` with all possible supertypes (I guess...)
replaced with 
```
CanBeSubclassDeterminer.isAssignable(boxedGivenClass, candidate)
```
that checks  `boxedGivenClass` with exact given supertype. Aaaand that possibly collides with the `java.math.BigDecimal is quite often returned as a wrapper for all kind of numbers`.

Reply:
Yeah, apart from this line the logic is exactly the same. To clarify, I think the current staging logic is buggy. It only checks whether a givenType CAN be converted, not if the conversion is to the givenSuperclass. Apart from that I've only included a fallback in case lang3 throws a false false, because the lib gives errors when it comes to decimal java types. In general the program execution flow is the same as intended before.

Reply:
I have moved this logic to a separate method and restored the previous logic in the handleNumberConversions method